// @web3auth/node-sdk/dist/analytics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ANALYTICS_EVENTS = exports.SegmentAnalytics = exports.SDK_VERSION = exports.SDK_TYPE = void 0;
const analytics_node_1 = require("@segment/analytics-node");
const no_modal_1 = require("@web3auth/no-modal");
// Import package.json for version info
// eslint-disable-next-line @typescript-eslint/no-require-imports
const packageJson = require("../package.json");
const SEGMENT_WRITE_KEY = "f6LbNqCeVRf512ggdME4b6CyflhF1tsX";
exports.SDK_TYPE = "node";
exports.SDK_VERSION = packageJson.version;
class SegmentAnalytics {
    segment;
    globalProperties = {};
    enabled = true;
    userId = null;
    init() {
        if (this.isSkipped()) {
            return;
        }
        if (this.segment) {
            throw new Error("Analytics already initialized");
        }
        this.segment = new analytics_node_1.Analytics({ writeKey: SEGMENT_WRITE_KEY });
    }
    enable() {
        this.enabled = true;
    }
    disable() {
        this.enabled = false;
    }
    setGlobalProperties(properties) {
        this.globalProperties = { ...this.globalProperties, ...properties };
    }
    async identify(userId, traits) {
        if (!this.enabled)
            return;
        if (this.isSkipped())
            return;
        try {
            this.userId = userId;
            return this.getSegment().identify({ userId, traits });
        }
        catch (error) {
            no_modal_1.log.error(`Failed to identify user ${userId} in analytics`, error);
        }
    }
    async track(event, properties) {
        if (!this.enabled)
            return;
        if (this.isSkipped())
            return;
        try {
            return this.getSegment().track({ userId: this.userId, event, properties: { ...properties, ...this.globalProperties } });
        }
        catch (error) {
            no_modal_1.log.error(`Failed to track event ${event}`, error);
        }
    }
    getSegment() {
        if (!this.segment) {
            no_modal_1.log.error("Analytics not initialized. Call Analytics.init() first.");
            throw new Error("Analytics not initialized. Call Analytics.init() first.");
        }
        return this.segment;
    }
    isSkipped() {
        // skip if the environment is not production
        if (process.env.NODE_ENV !== "production") {
            return true;
        }
        return false;
    }
}
exports.SegmentAnalytics = SegmentAnalytics;
exports.ANALYTICS_EVENTS = {
    // SDK Initialization
    SDK_INITIALIZATION_COMPLETED: "SDK Initialization Completed",
    SDK_INITIALIZATION_FAILED: "SDK Initialization Failed",
    // Connection
    CONNECTION_STARTED: "Connection Started",
    CONNECTION_COMPLETED: "Connection Completed",
    CONNECTION_FAILED: "Connection Failed",
};
